### ISSUE TITLE
3mf cannot be opened after a save

### ISSUE THREAD

**Issue by robustini** [supermerill/SuperSlicer#997]
### Version
2.3.55

### Operating system type + version
Windows 10 Pro

### Behavior
I have saved a project and can't open it anymore.
The annoying thing is that once I have established this I have to start saving projects with different revisions each time. 

![immagine](https://user-images.githubusercontent.com/3864712/110086024-2555a880-7d92-11eb-877f-23c870debaf9.png)

#### Project File (.3MF) where problem occurs

[Coda_Fusoliera.zip](https://github.com/supermerill/SuperSlicer/files/6089023/Coda_Fusoliera.zip)

**Comment by formadi** [comment]
me too

**Comment by supermerill** [comment]
Your project file has defined a fill apttern for every model, but some are "'empty" (it's in the file SuperSlicer_model.config in metadata inside the zip that is the .3mf)
I tried to replicate, to create a file with that kind of bug but i can't, every time a good name is written in these fields.

What did you do to get this bad save file?

**Comment by robustini** [comment]
@supermerill i just saved the file on my nvme, like i do with any other application.
I've never had this type of problem with other programs, so I think the corruption is due to a SuperSlicer issue.

**Comment by robustini** [comment]
> 
> 
> Your project file has defined a fill apttern for every model, but some are "'empty" (it's in the file SuperSlicer_model.config in metadata inside the zip that is the .3mf)
> I tried to replicate, to create a file with that kind of bug but i can't, every time a good name is written in these fields.
> 
> What did you do to get this bad save file?

I tried manually to correct the "fill pattern" in the xml file, since I remember they were all "rectilinear" and now the project file opens without any problem.
But I perfectly remember selecting "rectilinear" when I created the project.

**Comment by supermerill** [comment]
Of course, the slicer is the problem, but I don't know why it does this, and so if I want to correct the issue, I have to replicate the problem to understand why it's doing that.

So please try to reproduce the problem, and write here the steps.

**Comment by robustini** [comment]
> 
> 
> Of course, the slicer is the problem, but I don't know why it does this, and so if I want to correct the issue, I have to replicate the problem to understand why it's doing that.
> 
> So please try to reproduce the problem, and write here the steps.

Step to reproduce:

1) add modifier box with rectilinear infill to the STL
2) copy and paste this box and move then to another place
3) copy and paste the latest modifier box to another place
4) save

From what I see as long as the "modifier boxes" are only 2 it does not happen, from the third onwards yes.

**Comment by supermerill** [comment]
> * add modifier box with rectilinear infill to the STL
> * copy and paste this box and move them to another place
> * copy and paste the latest modifier box to another place
> * save

even with that, I can't reproduce the save bug.
what's your version? (in help-> about)

**Comment by robustini** [comment]
@supermerill it is very strange that you cannot replicate the problem.
I just need to add some "modifier" and every now and then copy them, collate them and then move them.
It also happened to me this morning with the version compiled by the current master, and even editing the xml manually now I can not solve the problem.

[Ala_SX_MID.zip](https://github.com/supermerill/SuperSlicer/files/6376365/Ala_SX_MID.zip)


![immagine](https://user-images.githubusercontent.com/3864712/116069784-14d3f700-a68c-11eb-92b3-f046f3fe2856.png)

**Comment by supermerill** [comment]
@robustini

Can you use [this test build](https://mega.nz/file/h0R0kZRY#th9MkRSWEg1GJwXY2e6nrbUZgoSIwt2fVZYjnDgj5Lk) (using the superslicer_console.exe)
When you will be exporting a buggy .3mf, some lines should appear in the console, can you write them here?

**Comment by robustini** [comment]
@supermerill 

I am trying to use it but everything is terribly slow, is this normal? 
Indeed I would say that at least here it is unusable.
To do the slicing where it used to take 20 seconds now takes over 5 minutes.
During any operation I don't see anything in the console. 

![immagine](https://user-images.githubusercontent.com/3864712/116193937-7057c080-a730-11eb-84fc-c28f40791e81.png)

**Comment by supermerill** [comment]
> I am trying to use it but everything is terribly slow, is this normal?

yes, it's debug build

> During any operation I don't see anything in the console.

it's only writing when saving the project, as the problem is here.

**Comment by robustini** [comment]
Se the latest lines...

![immagine](https://user-images.githubusercontent.com/3864712/116225341-00f2c880-a752-11eb-8196-0ac77a016c62.png)

**Comment by supermerill** [comment]
[here is an other test build](https://mega.nz/file/xs5UXLST#CamRwRNnykcIZblSSou6Tz2F9ot5xC1w9XTyfo3iN80) with a possible fix.
So if it creates a problema gain, it will output more logs, and there is a possibility that it doesn't crash.

**Comment by robustini** [comment]
Same issue at the first attempt.

![immagine](https://user-images.githubusercontent.com/3864712/116280109-a2950c80-a788-11eb-9891-f7182f6556d1.png)

**Comment by supermerill** [comment]
Erf, sadely it just confirm that this bugs is very weird.

I cleaned up a bit some definitions that weren't up to the latest c++11 thread-safe standard.
Also added more logs to try to pinpoint exactly what's going on.

[here the build to test](https://mega.nz/file/E8xSGQII#yHBr99NlmHY99CQ51BGf4XSekXZFIYmhLRSrXeN0XzM)

I expect at least 1 or 2 tests to follow.

**Comment by robustini** [comment]
Mouse right click on the piece and this appeared.
Obviously then the app closes.
So I tried again to do the same things, created a shape box, rescaled, added some modifier, to the third added it always does this.

![immagine](https://user-images.githubusercontent.com/3864712/116355223-d65c4a80-a7f9-11eb-90b2-a67ee36b6eaf.png)


![immagine](https://user-images.githubusercontent.com/3864712/116355203-cfcdd300-a7f9-11eb-8679-87e527ea06e3.png)

**Comment by supermerill** [comment]
this assert error has nothing to do with our bug, it's only because it's a debug build and some of the gui code isn't that robust.
can you retest some times get back to the usual bug?
If you can't get it back, maybe that means that the bug is resolved .

**Comment by supermerill** [comment]
Bad assert, don't let the right selection on a setting, click on the objet/a modifier before right-clic.
note that you can also right-clic on the paper at the right of your object.

>  i see now there's a fix in the Prusa Slicer master

where? can't spot it.

**Comment by robustini** [comment]
> 
> 
> Bad assert, don't let the right selection on a setting, click on the objet/a modifier before right-clic.
> note that you can also right-clic on the paper at the right of your object.

You're right!
I added 30 modifiers and it never gave a problem, do you think it's ok now?  
 
> where? can't spot it.

I don't know, but [here](https://github.com/prusa3d/PrusaSlicer/issues/5988) they say that in the 2.4 master the problem has been solved.

**Comment by supermerill** [comment]
>I don't know, but here they say that in the 2.4 master the problem has been solved.
yeah!
I'm going to remove the log and let you test it a bit.
Then I'm going to submit the fix also to PS.

> I don't know, but here they say that in the 2.4 master the problem has been solved.

that's for infill, not perimeters.

**Comment by supermerill** [comment]
[here is the release candidate}(https://mega.nz/file/5sgikbgA#cfUfDZfxL2fdqiBcz1lIzW1yOTJlklOG7wXAGGi3Mqg)
there are no logs.

**Comment by robustini** [comment]
Unfortunately I am sorry to inform you that this is still happening.

![immagine](https://user-images.githubusercontent.com/3864712/153614124-e1e77754-6b04-4bdb-ade5-6102125dfa93.png)

**Comment by robustini** [comment]
@supermerill I don't know if it is a similar problem but now it is perhaps worse.
Try opening this saved with today's nightly.

![immagine](https://user-images.githubusercontent.com/3864712/158054148-a0d84cb4-5ecc-4c53-a01d-c82ba9feedff.png)


[FUSO04.zip](https://github.com/supermerill/SuperSlicer/files/8239296/FUSO04.zip)
